packages/clay-multi-select/stories/index.tsx

```diff
@@ -235,6 +235,30 @@ storiesOf('Components|ClayMultiSelect', module)
 			/>
 		</>
 	))
+	.add('async', () => {
+		const [value, setValue] = React.useState('');
+		const [items, setItems] = React.useState([]);
+
+		function asyncData(query: string): Promise<Array<any>> {
+			return new Promise((res) => {
+				setTimeout(() => {
+					res(sourceItems.filter((item) => item.label.match(query)));
+				}, 2000);
+			});
+		}
+
+		return (
+			<ClayMultiSelect
+				inputName="myInput"
+				inputValue={value}
+				items={items}
+				onChange={setValue}
+				onItemsChange={setItems}
+				sourceItems={asyncData}
+				spritemap={spritemap}
+			/>
+		);
```

Comment:
I had this in mind for `sourceItems` to accept only one `Promise` and we hope it resolves to get the data I think it would be easier than accepting a function that returns a `Promise`, something like this:

```jsx
<ClayMultiSelect
	inputName="myInput"
	inputValue={value}
	items={items}
	onChange={setValue}
	onItemsChange={setItems}
	sourceItems={new Promise(...)}
	spritemap={spritemap}
/>
```

Reply:
@matuzalemsteles passing a Promise as a prop or argument seems odd to me for some reason. I feel like I am used to seeing functions that return a promise in similar APIs, I recall . This way we can also pass the query directly to the function. 

If we had a util function for calling the api it might look something like this, which seems strange to invoke on render rather than on the value change. At this level, you might as well hoist the async logic out and just us an `isLoading` prop.
```jsx
function myapp() {
  const itemFetcher = () => {
    // some business logic for wiring up call.
    return fetch(`/some/api?q=${value}`);
  };

  return (
    <ClayMultiSelect
      inputValue={value}
      items={items}
      onChange={setValue}
      onItemsChange={setItems}
      sourceItems={itemFetcher()}
    />
  );
}
```

Reply:
I could be looking at this wrong though. Can you think of other examples where we pass a promise directly?

Reply:
One argument in favor of a function is that detecting it is a tiny bit easier (eg. `typeof blah === 'function'`) compared with identifying a "thenable".

But with promises, they are generally pretty annoying to work with because they can't be cancelled, and they are opaque (ie. you can't ask them if they are resolved or not — this is why there are so many `PublicPromise` abstractions out there, which are wrappers around promises that add methods for introspecting their internal state).

Having said that, a function that returns a promise isn't any better in this regard; you still get a promise in the end, with all of its downsides.

One thing I notice about your implementation, @bryceosterhaus, is that you have a `then()` but no `catch()`; let's that add that to the list of things to hate about promises — that it is far too easy to overlook error-handling in them.

Just thinking out loud here, but a "best of both worlds" approach might be to start using `async`/`await` in Clay, although it does have obvious costs (eg. because Clay is used on older branches, we can't assume native browser support and would have to transpile). That way you could accept an async function, or a function that returns a promise, from the calling code, an inside Clay, we would at least have the somewhat more pleasing:

```ts
try {
	const asyncItems = await sourceItems(inputValue);

	setInternalSourceItems(asyncItems);

	filterAndSet(asyncItems);
}
catch {
	// Do something graceful; I don't actually know what...
}
finally {
	setIsLoading(false);
}
```

Ah, and one final thought... all those `set` calls run the risk of firing after the component unmounts, triggering an error. Any async function in a React components needs to be protected by some kind of "is mounted" check, like we do in DXP with [the `useIsMounted()` hook](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/hooks/useIsMounted.ts); not sure if we already have something like that somewhere in Clay.

Anyway, food for thought.

Reply:
Well, the idea of receiving only `Promise` instead of invoking a function that returns a `Promise` is that we don’t have the responsibility to call the function and we leave the responsibility to the developer who can call the function at any time he wants so he can add his own rules although he can do the same in this implementation we limit this so that `sourceItems` will be called whenever there is a change in `inputValue`, for example, the use of `useResource` he deals with the responsibility to revalidate the request when a `variable` changes, that is, we allow this to be composed by the developer and he can arrive at different business rules, where he can revalidate the request to any change in `inputValue` or just invoke just once.

The [Autocomplete example in the Storybook](https://storybook.clayui.com/?path=/story/components-clayautocomplete--w-async-data) shows some of these benefits, we add loading only when a request happens, when the user tries to search for a request that is in the cache we will return from the cache, this brings some beneficious, which we would not be able to do here and if to support we would have to do more things. I'm just thinking of removing the responsibility of this from the component.

> Just thinking out loud here, but a "best of both worlds" approach might be to start using async/await in Clay

Yeah, I think it would be worth trying to start using `async` / `await` but it really brings the responsibility for whoever uses it, dealing with transpilation.

> Any async function in a React components needs to be protected by some kind of "is mounted" check, like we do in DXP with the useIsMounted() hook...

Yeah, I think it's super valid to add an implementation of this to `shared`, we usually don't deal with asynchronous operations on other components except the implementation of `useResource` that has this responsibility, it would be good to just keep there...

Reply:
Thank you very much guys!
I just visually link the loading spinner to the data fetch action so my assumption is that the code which manages both the things should be in the same place.

My only (really small point) is that something like this: 
```
getData(query).then(updateSourceItems()) 
```
**looks** a bit more predictable than:

```
updateSourceItems(getData(query))
```
since dealing with unsolved promises It's pretty rare to me: where I don't use a `.then` they are wrapped in some `Promise` method like all, allSettled, etc...

That being said, this PR solves the problem I asked to be solved so to me is totally ok :) 
The `isLoading` would just make me slightly happier but for that I can get a dog :P

Reply:
Well I'm almost accepting the idea of just having `isLoading` or another name to show `LoadingIndicator`  🙈